# Pandas Notları

In [18]:
import pandas as pd 

In [22]:
df = pd.read_csv("AdvancedHousePricing.csv", index_col="Id")

In [23]:
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                  ...                                     
1          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
2          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   
3          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
4          Lvl    AllPub    Corner  ...        0    NaN   NaN         NaN   
5          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   

   MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                             
1        0      2    2008        WD         Normal     208500  
2        0      5    2007        WD         Normal     181500  
3        0      9    2008        WD         Normal     223500  
4        0      2    2006        WD        Abnorml     140000  
5        0     12    2008        WD         Normal     250000  

[5 rows x 80 columns]

### DataFrames practice

##### Activities

How many elements does this dataframe have?

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1460 entries, 1 to 1460
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MSSubClass     1460 non-null   int64  
 1   MSZoning       1460 non-null   object 
 2   LotFrontage    1201 non-null   float64
 3   LotArea        1460 non-null   int64  
 4   Street         1460 non-null   object 
 5   Alley          91 non-null     object 
 6   LotShape       1460 non-null   object 
 7   LandContour    1460 non-null   object 
 8   Utilities      1460 non-null   object 
 9   LotConfig      1460 non-null   object 
 10  LandSlope      1460 non-null   object 
 11  Neighborhood   1460 non-null   object 
 12  Condition1     1460 non-null   object 
 13  Condition2     1460 non-null   object 
 14  BldgType       1460 non-null   object 
 15  HouseStyle     1460 non-null   object 
 16  OverallQual    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  YearBuilt    

In [26]:
df.shape

(1460, 80)

What is the 'MSSubClass' value of the 56(index)?

In [ ]:
#df.loc[INDEX, COLUMNS]

In [30]:
df.loc[56] #It is entered the index number.
# Don't forget to use "" if the index is string.
# loc brings the row values in that index.

MSSubClass           20
MSZoning             RL
LotFrontage       100.0
LotArea           10175
Street             Pave
                  ...  
MoSold                7
YrSold             2008
SaleType             WD
SaleCondition    Normal
SalePrice        180500
Name: 56, Length: 80, dtype: object

What is the highest possible 'LotArea' value of an Id(index)?

In [32]:
df["LotArea"].max()

215245

In [ ]:
df.max()
#the max statistical methoh computed for the whole dataframe 
# It is valid only dataframe columns that have all numerical value 

In [35]:
df.describe()
# summary statistics of all numeric columns

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond  \
count  1460.000000  1201.000000    1460.000000  1460.000000  1460.000000   
mean     56.897260    70.049958   10516.828082     6.099315     5.575342   
std      42.300571    24.284752    9981.264932     1.382997     1.112799   
min      20.000000    21.000000    1300.000000     1.000000     1.000000   
25%      20.000000    59.000000    7553.500000     5.000000     5.000000   
50%      50.000000    69.000000    9478.500000     6.000000     5.000000   
75%      70.000000    80.000000   11601.500000     7.000000     6.000000   
max     190.000000   313.000000  215245.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2  ...  \
count  1460.000000   1460.000000  1452.000000  1460.000000  1460.000000  ...   
mean   1971.267808   1984.865753   103.685262   443.639726    46.549315  ...   
std      30.202904     20.645407   181.066207   456.098091   161.319273  ...   
min    1872.000000   1950.000000     0.000000     0.000000     0.000000  ...   
25%    1954.000000   1967.000000     0.000000     0.000000     0.000000  ...   
50%    1973.000000   1994.000000     0.000000   383.500000     0.000000  ...   
75%    2000.000000   2004.000000   166.000000   712.250000     0.000000  ...   
max    2010.000000   2010.000000  1600.000000  5644.000000  1474.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 37 columns]

Which of the following 'Id' have a 'MSSubClass' of 60 and a 'LotArea' of 11250?

In [36]:
# It is put possible Ids to observe and pick the wanted one.
df.loc[[
    2,
    45,
    68,
    21,
    3
]]


MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
2           20       RL         80.0     9600   Pave   NaN      Reg   
45          20       RL         70.0     7945   Pave   NaN      Reg   
68          20       RL         72.0    10665   Pave   NaN      IR1   
21          60       RL        101.0    14215   Pave   NaN      IR1   
3           60       RL         68.0    11250   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                  ...                                     
2          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   
45         Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
68         Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
21         Lvl    AllPub    Corner  ...        0    NaN   NaN         NaN   
3          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   

   MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                             
2        0      5    2007        WD         Normal     181500  
45       0      5    2006        WD         Normal     141000  
68       0      6    2007        WD         Normal     226000  
21       0     11    2006       New        Partial     325300  
3        0      9    2008        WD         Normal     223500  

[5 rows x 80 columns]

Winner 'Id' is 3. Yay!!!

In [37]:
df.loc[[
    2,
    45,
    68,
    21,
    3
], "LotArea"] #In this way, it can be seen particular column values for multiple indexes.

Id
2      9600
45     7945
68    10665
21    14215
3     11250
Name: LotArea, dtype: int64

What is the highest possible length of a 'LotFrontage'?

In [38]:
df.describe()

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond  \
count  1460.000000  1201.000000    1460.000000  1460.000000  1460.000000   
mean     56.897260    70.049958   10516.828082     6.099315     5.575342   
std      42.300571    24.284752    9981.264932     1.382997     1.112799   
min      20.000000    21.000000    1300.000000     1.000000     1.000000   
25%      20.000000    59.000000    7553.500000     5.000000     5.000000   
50%      50.000000    69.000000    9478.500000     6.000000     5.000000   
75%      70.000000    80.000000   11601.500000     7.000000     6.000000   
max     190.000000   313.000000  215245.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2  ...  \
count  1460.000000   1460.000000  1452.000000  1460.000000  1460.000000  ...   
mean   1971.267808   1984.865753   103.685262   443.639726    46.549315  ...   
std      30.202904     20.645407   181.066207   456.098091   161.319273  ...   
min    1872.000000   1950.000000     0.000000     0.000000     0.000000  ...   
25%    1954.000000   1967.000000     0.000000     0.000000     0.000000  ...   
50%    1973.000000   1994.000000     0.000000   383.500000     0.000000  ...   
75%    2000.000000   2004.000000   166.000000   712.250000     0.000000  ...   
max    2010.000000   2010.000000  1600.000000  5644.000000  1474.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 37 columns]

What is the 'LotFrontage' with the value of 313.0?

In [41]:
df.sort_values(by=["LotFrontage"], ascending=False)
# We sorted from largest to smallest by using ascending=False in a numerical column of our choice.

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1299          60       RL        313.0    63887   Pave   NaN      IR3   
935           20       RL        313.0    27650   Pave   NaN      IR2   
1128          20       RL        182.0    14572   Pave   NaN      IR3   
198           75       RL        174.0    25419   Pave   NaN      Reg   
232           60       RL        174.0    15138   Pave   NaN      IR1   
...          ...      ...          ...      ...    ...   ...      ...   
1430          20       RL          NaN    12546   Pave   NaN      IR1   
1432         120       RL          NaN     4928   Pave   NaN      IR1   
1442         120       RM          NaN     4426   Pave   NaN      Reg   
1444          30       RL          NaN     8854   Pave   NaN      Reg   
1447          20       RL          NaN    26142   Pave   NaN      IR1   

     LandContour Utilities LotConfig  ... PoolArea PoolQC  Fence MiscFeature  \
Id                                    ...                                      
1299         Bnk    AllPub    Corner  ...      480     Gd    NaN         NaN   
935          HLS    AllPub    Inside  ...        0    NaN    NaN         NaN   
1128         Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
198          Lvl    AllPub    Corner  ...      512     Ex  GdPrv         NaN   
232          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
...          ...       ...       ...  ...      ...    ...    ...         ...   
1430         Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
1432         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1442         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1444         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1447         Lvl    AllPub   CulDSac  ...        0    NaN    NaN         NaN   

     MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                               
1299       0      1    2008       New        Partial     160000  
935        0     11    2008        WD         Normal     242000  
1128       0     11    2007        WD         Family     259000  
198        0      3    2006        WD        Abnorml     235000  
232        0      7    2009        WD         Normal     403000  
...      ...    ...     ...       ...            ...        ...  
1430       0      4    2007        WD         Normal     182900  
1432       0     10    2009        WD         Normal     143750  
1442       0      5    2008        WD         Normal     149300  
1444       0      5    2009        WD         Normal     121000  
1447       0      4    2010        WD         Normal     157900  

[1460 rows x 80 columns]

In [ ]:
#It does not work for this datadframe??
for row in df:
    if row["MSSubClass"] == 20:
        print(row)

In [44]:
df.loc[df["LotFrontage"] == 313.0]
# This is the first idea of conditional selection in pandas.
# By using boolean values, we take up less space in memory compared to loops.

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
935           20       RL        313.0    27650   Pave   NaN      IR2   
1299          60       RL        313.0    63887   Pave   NaN      IR3   

     LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                    ...                                     
935          HLS    AllPub    Inside  ...        0    NaN   NaN         NaN   
1299         Bnk    AllPub    Corner  ...      480     Gd   NaN         NaN   

     MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                               
935        0     11    2008        WD         Normal     242000  
1299       0      1    2008       New        Partial     160000  

[2 rows x 80 columns]

In [45]:
(df["LotFrontage"] == 313.0)
# Actually, this operation returns a Boolean array.

Id
1       False
2       False
3       False
4       False
5       False
        ...  
1456    False
1457    False
1458    False
1459    False
1460    False
Name: LotFrontage, Length: 1460, dtype: bool

What is the most common 'LotFrontage' value?

In [52]:
df["LotFrontage"].describe()

count    1201.000000
mean       70.049958
std        24.284752
min        21.000000
25%        59.000000
50%        69.000000
75%        80.000000
max       313.000000
Name: LotFrontage, dtype: float64

In [54]:
df["LotFrontage"].mode()
# 60 is the most common value.

0    60.0
Name: LotFrontage, dtype: float64

In [55]:
df["LotFrontage"].value_counts().head()
# It gives value counts

LotFrontage
60.0    143
70.0     70
80.0     69
50.0     57
75.0     53
Name: count, dtype: int64

In [57]:
df.loc[df["LotFrontage"] == 60.0].head(10)
# It gives top 10 values.

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
4           70       RL         60.0     9550   Pave   NaN      IR1   
27          20       RL         60.0     7200   Pave   NaN      Reg   
30          30       RM         60.0     6324   Pave   NaN      IR1   
35         120       RL         60.0     7313   Pave   NaN      Reg   
55          80       RL         60.0     7134   Pave   NaN      Reg   
60          20       RL         60.0     7200   Pave   NaN      Reg   
62          75       RM         60.0     7200   Pave   NaN      Reg   
75          50       RM         60.0     5790   Pave   NaN      Reg   
80          50       RM         60.0    10440   Pave  Grvl      Reg   
90          20       RL         60.0     8070   Pave   NaN      Reg   

   LandContour Utilities LotConfig  ... PoolArea PoolQC  Fence MiscFeature  \
Id                                  ...                                      
4          Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
27         Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
30         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
35         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
55         Bnk    AllPub    Inside  ...        0    NaN  MnPrv         NaN   
60         Bnk    AllPub    Inside  ...        0    NaN  MnPrv         NaN   
62         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
75         Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
80         Lvl    AllPub    Corner  ...        0    NaN  MnPrv         NaN   
90         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   

   MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                             
4        0      2    2006        WD        Abnorml     140000  
27       0      5    2010        WD         Normal     134800  
30       0      5    2008        WD         Normal      68500  
35       0      8    2007        WD         Normal     277500  
55       0      2    2007        WD         Normal     130000  
60       0      1    2008        WD         Normal     124900  
62       0      3    2007        WD         Normal     101000  
75       0      5    2010        WD         Normal     107400  
80       0      5    2009        WD         Normal     110000  
90       0      8    2007        WD         Normal     123600  

[10 rows x 80 columns]

In [59]:
df.loc[df["LotFrontage"] == 60.0].sample(10)
# It gives randomly 10 values.

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
726           20       RL         60.0     6960   Pave   NaN      Reg   
1225          60       RL         60.0    15384   Pave   NaN      IR1   
1394         190       RM         60.0    10800   Pave  Pave      Reg   
1378          50       RL         60.0    10998   Pave  Grvl      Reg   
1093          50       RL         60.0     8400   Pave   NaN      Reg   
195           20       RL         60.0     7180   Pave   NaN      IR1   
90            20       RL         60.0     8070   Pave   NaN      Reg   
1399          50       RL         60.0     7200   Pave   NaN      Reg   
1293          70       RM         60.0     6600   Pave   NaN      Reg   
1145         190       RL         60.0    12180   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolArea PoolQC  Fence MiscFeature  \
Id                                    ...                                      
726          Lvl    AllPub    Inside  ...        0    NaN    NaN        Shed   
1225         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1394         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1378         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1093         Bnk    AllPub    Inside  ...        0    NaN    NaN         NaN   
195          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
90           Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1399         Lvl    AllPub    Inside  ...        0    NaN  GdPrv         NaN   
1293         Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
1145         Lvl    AllPub    Inside  ...        0    NaN  MnPrv         NaN   

     MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                               
726      500     11    2009        WD         Normal     120500  
1225       0      2    2008        WD         Normal     184000  
1394       0      4    2008        WD         Normal     163000  
1378       0      7    2009        WD         Normal     117000  
1093       0      6    2008        WD         Normal     136500  
195        0      5    2008        WD         Normal     127000  
90         0      8    2007        WD         Normal     123600  
1399       0      5    2009        WD         Normal     138000  
1293       0     12    2009        WD         Normal     107500  
1145       0      7    2010        WD         Normal      80000  

[10 rows x 80 columns]

What is the smallest 'MSSubClass' with the 'LotFrontage' value 60.0?

In [60]:
df.loc[df["LotFrontage"] == 60.0].sort_values(by="MSSubClass")

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1437          20       RL         60.0     9000   Pave   NaN      Reg   
27            20       RL         60.0     7200   Pave   NaN      Reg   
183           20       RL         60.0     9060   Pave   NaN      Reg   
157           20       RL         60.0     7200   Pave   NaN      Reg   
1315          20       RL         60.0     8190   Pave   NaN      Reg   
...          ...      ...          ...      ...    ...   ...      ...   
1267         190       RM         60.0    10120   Pave   NaN      IR1   
636          190       RH         60.0    10896   Pave  Pave      Reg   
714          190       RL         60.0     9873   Pave   NaN      Reg   
521          190       RL         60.0    10800   Pave  Grvl      Reg   
1417         190       RM         60.0    11340   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolArea PoolQC  Fence MiscFeature  \
Id                                    ...                                      
1437         Lvl    AllPub       FR2  ...        0    NaN   GdWo         NaN   
27           Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
183          Lvl    AllPub    Inside  ...        0    NaN  MnPrv         NaN   
157          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1315         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
...          ...       ...       ...  ...      ...    ...    ...         ...   
1267         Bnk    AllPub    Inside  ...        0    NaN  MnPrv         NaN   
636          Bnk    AllPub    Inside  ...        0    NaN    NaN         NaN   
714          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
521          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1417         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   

     MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                               
1437       0      5    2007        WD         Normal     120500  
27         0      5    2010        WD         Normal     134800  
183        0      6    2007        WD         Normal     120000  
157        0      6    2006        WD         Normal     109500  
1315       0     10    2007        WD         Normal     119000  
...      ...    ...     ...       ...            ...        ...  
1267       0      1    2007        WD         Normal     122000  
636        0      3    2007        WD        Abnorml     200000  
714        0      5    2006        WD         Normal     129000  
521        0      8    2008        WD         Normal     106250  
1417       0      4    2010        WD         Normal     122500  

[143 rows x 80 columns]

In [61]:
df.loc[df["LotFrontage"] == 60.0, "MSSubClass"].min()

20

In [62]:
df.loc[
    (df["LotFrontage"] == 60.0) &
    (df["MSSubClass"] == 20)
]

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
27            20       RL         60.0     7200   Pave   NaN      Reg   
60            20       RL         60.0     7200   Pave   NaN      Reg   
90            20       RL         60.0     8070   Pave   NaN      Reg   
91            20       RL         60.0     7200   Pave   NaN      Reg   
157           20       RL         60.0     7200   Pave   NaN      Reg   
183           20       RL         60.0     9060   Pave   NaN      Reg   
195           20       RL         60.0     7180   Pave   NaN      IR1   
282           20       FV         60.0     7200   Pave  Pave      Reg   
342           20       RH         60.0     8400   Pave   NaN      Reg   
382           20       FV         60.0     7200   Pave  Pave      Reg   
397           20       RL         60.0     7200   Pave   NaN      Reg   
463           20       RL         60.0     8281   Pave   NaN      IR1   
465           20       RL         60.0     8430   Pave   NaN      Reg   
533           20       RL         60.0     7200   Pave   NaN      Reg   
572           20       RL         60.0     7332   Pave   NaN      Reg   
593           20       RL         60.0     6600   Pave   NaN      Reg   
647           20       RL         60.0     7200   Pave   NaN      Reg   
689           20       RL         60.0     8089   Pave   NaN      Reg   
726           20       RL         60.0     6960   Pave   NaN      Reg   
822           20       RM         60.0     6000   Pave  Pave      Reg   
836           20       RL         60.0     9600   Pave   NaN      Reg   
871           20       RL         60.0     6600   Pave   NaN      Reg   
881           20       RL         60.0     7024   Pave   NaN      Reg   
951           20       RL         60.0     7200   Pave   NaN      Reg   
952           20       RH         60.0     7800   Pave   NaN      Reg   
1015          20       RL         60.0    11664   Pave   NaN      Reg   
1021          20       RL         60.0     7024   Pave   NaN      Reg   
1050          20       RL         60.0    11100   Pave   NaN      Reg   
1126          20       RL         60.0    10434   Pave   NaN      Reg   
1141          20       RL         60.0     7350   Pave   NaN      Reg   
1250          20       RL         60.0     7200   Pave   NaN      Reg   
1295          20       RL         60.0     8172   Pave   NaN      Reg   
1308          20       RL         60.0     8072   Pave   NaN      Reg   
1315          20       RL         60.0     8190   Pave   NaN      Reg   
1328          20       RL         60.0     6600   Pave   NaN      Reg   
1437          20       RL         60.0     9000   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolArea PoolQC  Fence MiscFeature  \
Id                                    ...                                      
27           Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
60           Bnk    AllPub    Inside  ...        0    NaN  MnPrv         NaN   
90           Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
91           Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
157          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
183          Lvl    AllPub    Inside  ...        0    NaN  MnPrv         NaN   
195          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
282          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
342          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
382          Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
397          Low    AllPub    Inside  ...        0    NaN    NaN         NaN   
463          Lvl    AllPub    Inside  ...        0    NaN   GdWo         NaN   
465          HLS    AllPub    Inside  ...        0    NaN    NaN         NaN   
533          Lvl    AllPub    I

Create a column 'Ratio', which represents the 'LotFrontage Ratio' of a house. LotFrontage/MSSubClass (I made it up, it does not mean anything.)

In [63]:
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                  ...                                     
1          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
2          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   
3          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
4          Lvl    AllPub    Corner  ...        0    NaN   NaN         NaN   
5          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   

   MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                             
1        0      2    2008        WD         Normal     208500  
2        0      5    2007        WD         Normal     181500  
3        0      9    2008        WD         Normal     223500  
4        0      2    2006        WD        Abnorml     140000  
5        0     12    2008        WD         Normal     250000  

[5 rows x 80 columns]

In [64]:
df["Ratio"] = df["LotFrontage"] / df["MSSubClass"]

In [65]:
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolQC Fence MiscFeature MiscVal  \
Id                                  ...                                    
1          Lvl    AllPub    Inside  ...    NaN   NaN         NaN       0   
2          Lvl    AllPub       FR2  ...    NaN   NaN         NaN       0   
3          Lvl    AllPub    Inside  ...    NaN   NaN         NaN       0   
4          Lvl    AllPub    Corner  ...    NaN   NaN         NaN       0   
5          Lvl    AllPub       FR2  ...    NaN   NaN         NaN       0   

   MoSold YrSold  SaleType  SaleCondition  SalePrice     Ratio  
Id                                                              
1       2   2008        WD         Normal     208500  1.083333  
2       5   2007        WD         Normal     181500  4.000000  
3       9   2008        WD         Normal     223500  1.133333  
4       2   2006        WD        Abnorml     140000  0.857143  
5      12   2008        WD         Normal     250000  1.400000  

[5 rows x 81 columns]

What is the maximum value of 'Ratio'?

In [66]:
df["Ratio"].max()

15.65

What "LotFrontage" value is the one with the highest 'Ratio'?

In [68]:
df.sort_values(by="Ratio", ascending=False).head()
# By default, values are sorting in ascending mode.

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
935           20       RL        313.0    27650   Pave   NaN      IR2   
1128          20       RL        182.0    14572   Pave   NaN      IR3   
314           20       RL        150.0   215245   Pave   NaN      IR3   
172           20       RL        141.0    31770   Pave   NaN      IR1   
278           20       RL        140.0    19138   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolQC Fence MiscFeature MiscVal  \
Id                                    ...                                    
935          HLS    AllPub    Inside  ...    NaN   NaN         NaN       0   
1128         Lvl    AllPub    Corner  ...    NaN   NaN         NaN       0   
314          Low    AllPub    Inside  ...    NaN   NaN         NaN       0   
172          Lvl    AllPub    Corner  ...    NaN   NaN         NaN       0   
278          Lvl    AllPub    Corner  ...    NaN   NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  Ratio  
Id                                                             
935      11   2008        WD         Normal     242000  15.65  
1128     11   2007        WD         Family     259000   9.10  
314       6   2009        WD         Normal     375000   7.50  
172       5   2010        WD         Normal     215000   7.05  
278       6   2010        WD         Normal     141000   7.00  

[5 rows x 81 columns]

In [69]:
df.loc[df["Ratio"] == df["Ratio"].max()]

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                     
935          20       RL        313.0    27650   Pave   NaN      IR2   

    LandContour Utilities LotConfig  ... PoolQC Fence MiscFeature MiscVal  \
Id                                   ...                                    
935         HLS    AllPub    Inside  ...    NaN   NaN         NaN       0   

    MoSold YrSold  SaleType  SaleCondition  SalePrice  Ratio  
Id                                                            
935     11   2008        WD         Normal     242000  15.65  

[1 rows x 81 columns]

How many words have a 'Ratio' of 4?

In [70]:
df["Ratio"].value_counts()
# Answer is 36 (4.000000)

Ratio
1.200000    43
3.500000    40
3.000000    37
4.000000    36
1.000000    35
            ..
0.750000     1
1.483333     1
6.700000     1
4.600000     1
5.450000     1
Name: count, Length: 309, dtype: int64

In [72]:
df.loc[df["Ratio"] == 4].head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                     
2            20       RL         80.0     9600   Pave   NaN      Reg   
84           20       RL         80.0     8892   Pave   NaN      IR1   
136          20       RL         80.0    10400   Pave   NaN      Reg   
174          20       RL         80.0    10197   Pave   NaN      IR1   
201          20       RM         80.0     8546   Pave   NaN      Reg   

    LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                   ...                                     
2           Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
84          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
136         Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
174         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
201         Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   

    MoSold YrSold  SaleType  SaleCondition  SalePrice  Ratio  
Id                                                            
2        5   2007        WD         Normal     181500    4.0  
84       7   2007       COD         Normal     126500    4.0  
136      5   2008        WD         Normal     174000    4.0  
174      6   2008       COD         Normal     163000    4.0  
201      3   2010        WD         Normal     140000    4.0  

[5 rows x 81 columns]

In [73]:
df.loc[df["Ratio"] == 4].shape

(36, 81)

In [74]:
#Query Method
df.query("Ratio == 4").shape

(36, 81)

What is the maximum value of all the 'LotFrontage' values with a 'Ratio' of 4?

In [77]:
df.query("Ratio == 4").sort_values(by="LotFrontage", ascending=False).head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1062          30  C (all)        120.0    18000   Grvl   NaN      Reg   
2             20       RL         80.0     9600   Pave   NaN      Reg   
136           20       RL         80.0    10400   Pave   NaN      Reg   
84            20       RL         80.0     8892   Pave   NaN      IR1   
174           20       RL         80.0    10197   Pave   NaN      IR1   

     LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                    ...                                     
1062         Low    AllPub    Inside  ...    NaN    NaN        Shed     560   
2            Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
136          Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
84           Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
174          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  Ratio  
Id                                                             
1062      8   2008     ConLD         Normal      81000    4.0  
2         5   2007        WD         Normal     181500    4.0  
136       5   2008        WD         Normal     174000    4.0  
84        7   2007       COD         Normal     126500    4.0  
174       6   2008       COD         Normal     163000    4.0  

[5 rows x 81 columns]

In [78]:
df.loc[df["Ratio"] == 4, "LotFrontage"].max()

120.0

Of those words with the 'LotFrontage' value is 60.0, what is the smallest 'MSSubClass' value found?

In [79]:
df.query("LotFrontage == 60.0")

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
4             70       RL         60.0     9550   Pave   NaN      IR1   
27            20       RL         60.0     7200   Pave   NaN      Reg   
30            30       RM         60.0     6324   Pave   NaN      IR1   
35           120       RL         60.0     7313   Pave   NaN      Reg   
55            80       RL         60.0     7134   Pave   NaN      Reg   
...          ...      ...          ...      ...    ...   ...      ...   
1429          30       RM         60.0     7200   Pave   NaN      Reg   
1431          60       RL         60.0    21930   Pave   NaN      IR3   
1433          30       RL         60.0    10800   Pave  Grvl      Reg   
1437          20       RL         60.0     9000   Pave   NaN      Reg   
1451          90       RL         60.0     9000   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                    ...                                     
4            Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
27           Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
30           Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
35           Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
55           Bnk    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
...          ...       ...       ...  ...    ...    ...         ...     ...   
1429         Lvl    AllPub    Corner  ...    NaN  MnPrv         NaN       0   
1431         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1433         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1437         Lvl    AllPub       FR2  ...    NaN   GdWo         NaN       0   
1451         Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice     Ratio  
Id                                                                
4         2   2006        WD        Abnorml     140000  0.857143  
27        5   2010        WD         Normal     134800  3.000000  
30        5   2008        WD         Normal      68500  2.000000  
35        8   2007        WD         Normal     277500  0.500000  
55        2   2007        WD         Normal     130000  0.750000  
...     ...    ...       ...            ...        ...       ...  
1429      2   2010        WD        Abnorml     119000  2.000000  
1431      7   2006        WD         Normal     192140  1.000000  
1433      8   2007        WD         Normal      64500  2.000000  
1437      5   2007        WD         Normal     120500  3.000000  
1451      9   2009        WD         Normal     136000  0.666667  

[143 rows x 81 columns]

In [80]:
df.query("LotFrontage == 60.0").sort_values(by="MSSubClass")

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1437          20       RL         60.0     9000   Pave   NaN      Reg   
27            20       RL         60.0     7200   Pave   NaN      Reg   
183           20       RL         60.0     9060   Pave   NaN      Reg   
157           20       RL         60.0     7200   Pave   NaN      Reg   
1315          20       RL         60.0     8190   Pave   NaN      Reg   
...          ...      ...          ...      ...    ...   ...      ...   
1267         190       RM         60.0    10120   Pave   NaN      IR1   
636          190       RH         60.0    10896   Pave  Pave      Reg   
714          190       RL         60.0     9873   Pave   NaN      Reg   
521          190       RL         60.0    10800   Pave  Grvl      Reg   
1417         190       RM         60.0    11340   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                    ...                                     
1437         Lvl    AllPub       FR2  ...    NaN   GdWo         NaN       0   
27           Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
183          Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
157          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1315         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
...          ...       ...       ...  ...    ...    ...         ...     ...   
1267         Bnk    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
636          Bnk    AllPub    Inside  ...    NaN    NaN         NaN       0   
714          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
521          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1417         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice     Ratio  
Id                                                                
1437      5   2007        WD         Normal     120500  3.000000  
27        5   2010        WD         Normal     134800  3.000000  
183       6   2007        WD         Normal     120000  3.000000  
157       6   2006        WD         Normal     109500  3.000000  
1315     10   2007        WD         Normal     119000  3.000000  
...     ...    ...       ...            ...        ...       ...  
1267      1   2007        WD         Normal     122000  0.315789  
636       3   2007        WD        Abnorml     200000  0.315789  
714       5   2006        WD         Normal     129000  0.315789  
521       8   2008        WD         Normal     106250  0.315789  
1417      4   2010        WD         Normal     122500  0.315789  

[143 rows x 81 columns]

In [81]:
df["MSSubClass"].describe()

count    1460.000000
mean       56.897260
std        42.300571
min        20.000000
25%        20.000000
50%        50.000000
75%        70.000000
max       190.000000
Name: MSSubClass, dtype: float64

*Find the all 'Id's with 'MSSubClass'> avg MSSubClass

In [82]:
MSSubClass_mean = df["MSSubClass"].mean()
MSSubClass_mean

56.897260273972606

In [ ]:
df.query("'MSSubClass' > 56") # dont forget inner quote!!!
# IT does not work for this data

In [ ]:
df.query("'MSSubClass' > @MSSubClass_mean") # using @ is essential.
# It does not work either.

33'26''

### Filtering and sorting with Pokemon data